Allow sorting schema properties by name

bodograumann opened the pull request:
Currently it is not possible to sort properties, schemas or responses by name, because any order defined for them is applied to the nested objects instead.
In order to distinguish between direct ordering of an object and nested ordering of its keys, I have introduced a new notation `properties[*]` here akin to [JSONPath](http://jsonpath.com/) as used by [Spectral](https://docs.stoplight.io/docs/spectral/01baf06bdd05a-create-a-ruleset).
Ideally the sort order would use complete path expressions, but that will take much more effort.

I know this is a breaking change for anyone using a custom sort order, but I think it is worthwhile.
The current implementation is limited and I found it difficult to understand how a custom sortFile even works. First I used trial-and-error to see what can and can't be sorted and then I looked at the source code.
E.g. the default sort order defines `content: []`, which means sort anything under a `content` key alphabetically and it has `properties: […]`, which means sort the fields of all objects nested inside the `properties` key.

- Simplify branching in sort function
- Explicitly distinguish between direct and nested sorting
- Test sorting schema properties by name

ed-randall-blk commented:
Please could you clarify, how do you configure it so that:
1. Your parameters are sorted by arg name
2. The parameter property names are sorted as expected
I tried
```
  "parameters[*]": ["name", "description", "in", "type", "format", "required", "default", "maximum", "minimum", "schema"],
  "parameters": ["name"]
```
but it doesn't seem to be sorting the parameters by name

thim81 commented:
@bodograumann 

Thanks for the PR, I still have to find time to review the code and the impact of the breaking changes.
I understand the need to sort properties, schemas or responses by name, is the additional test https://github.com/thim81/openapi-format/pull/78/files#diff-2ed84511e150b344f99bcb395066fc8fd3b8a9e7580406e99a78ace4ee63fa85 a good example for me to fully understand the expected behaviour?

thim81 commented:
@ed-randall-blk @bodograumann 

Following up on the progress and the findings.

1. Is the PR still needed or is using an [] the solution?  If yes, I'll document this in the readme for future users.
Since I'm cautious of introducing the new [*] syntax because it feels more complex to configure and would potentially break existing configurations.
2. The "openapi" property is needed to [detect the "ROOT" level](https://github.com/thim81/openapi-format/blob/eb83952b5b89ea5099c849ca2a9cc95006c416e8/openapi-format.js#L101C2-L101C2) to apply the sorting of the [root elements](https://github.com/thim81/openapi-format/blob/eb83952b5b89ea5099c849ca2a9cc95006c416e8/defaultSort.json#L2), plus I assumed that the "openapi" would always be present in OpenAPI documents since openapi-format was supporting: OpenAPI 3.0 & 3.1. 
Is it compatible with "swagger": "2.0"?

bodograumann commented:
Yes this PR is definitely still needed to be able to sort these things. 
E.g. `properties` is an array of objects.
We want to sort this array in some way; according to a key in the objects.
Then we _also_ want to sort the keys in each property object in a fixed order.
These two operations need separate configuration and without this PR it is not possible to sort the array at all.

I tried to keep the breakage to a minimum.
Another approach would be to define a completely separate new sorting method, so users could opt-in to the new, more powerful approach, while existing use cases would not be affected. This would also allow us to do more drastic redesigns of the sorting configuration.
That would require a bit more effort in terms of duplicating tests though.

What do you think, @thim81 ?

ed-randall-blk commented:
The "swagger.json" (swagger 2.0) document that I'm working with seems fully compatible with the tool (for sorting purposes, at least) other than that it has "swagger: 2.0" instead of "openapi: 3.0" at the top;   I tried to update the python to add both tags - but this fails the swagger validation :
```
- .openapi in body is a forbidden property
```
So it would be useful if openapi-format would recognise either tag at the root.

thim81 commented:
hi @bodograumann & @ed-randall-blk 

Re-opening the discussion.
I had already foreseen a mechanism to sort by using the  `--sortComponentsFile`  option.
By defining which elements that should be sorted, the CLI will sort the elements alphabetically.

Have a look at the tests:
- https://github.com/thim81/openapi-format/tree/main/test/json-sort-request-params
- https://github.com/thim81/openapi-format/tree/main/test/yaml-sort-components

Given that this option, will this solve your use-case?

bodograumann commented:
I'll look into it, @thim81 , but it is really confusing when `sortComponentsFile` sorts thing that are not within the components key.

bodograumann commented:
Let me be a bit more in-depth.
Here are the things that should be sortable imho, together with some examples:

1. Arrays of objects alphabetically by a value of a specific key in the children
    - parameters of a rest operation
    - top-level `server` and others
2. Arrays of strings alphabetically by their values
    - enum options
    - required fields
3. Objects with a fixed set of keys by listing the key order
    - top level `info` and others
    - any path (keys are the rest verbs like `get:` and `parameters:`)
    - any endpoint definition below the rest verbs
    - all kind of definitions that you would find under `components`, but also inline under the paths
4. Maps with a variable key alphabetically by the key
    - properties of on object schema
    - responses with the response code as key
    - responseBodies with the content-type of key
    - all children of `components`


Going over the current `openapi-format.js` in main:

- [Lines 53 - 62](https://github.com/thim81/openapi-format/blob/0f6a408355e6bd28c191cf3b7ea3e4d04c2365e7/openapi-format.js#L53-L62):
    Any child of `components` can be sorted according to (4) above, when defined in `sortComponents`. This does not include any of the examples in (4) outside of `components` though.
- [Lines 63 - 71](https://github.com/thim81/openapi-format/blob/0f6a408355e6bd28c191cf3b7ea3e4d04c2365e7/openapi-format.js#L63-L71):
    Any array outside of `components` can be sorted according to (1), when defined in `sortComponents`, but only by the `name` key of the items, which does not always exist (e.g. `servers`). Furthermore there can be list of `parameters` nested inside reusable `components.pathItems`. So again (1) is not fully covered.
- [Lines 72 - 74](https://github.com/thim81/openapi-format/blob/0f6a408355e6bd28c191cf3b7ea3e4d04c2365e7/openapi-format.js#L72-L74):
    Here we start sorting by `sort` instead of `sortComponents`, when the current key is configured there.
- [Lines 75 - 88](https://github.com/thim81/openapi-format/blob/0f6a408355e6bd28c191cf3b7ea3e4d04c2365e7/openapi-format.js#L75-L88):
    Objects are sorted here by an order on the keys as per (3), but only when they are nested inside an array. This is so that we can associate them with the identifier in the configuration.
- [Lines 89 - 98](https://github.com/thim81/openapi-format/blob/0f6a408355e6bd28c191cf3b7ea3e4d04c2365e7/openapi-format.js#L89-L98):
    Sort the children of `responses`, `schemas` and `properties` according to (3). So here the parent is not an array as directly before, but an object. This fails to cover the children in `paths` for example.
- [Lines 99 - 110](https://github.com/thim81/openapi-format/blob/0f6a408355e6bd28c191cf3b7ea3e4d04c2365e7/openapi-format.js#L99-L110):
    This also covers (3), but at this point associating the order not via the key of the parent node, but by the key of the object itself.

All in all I must say that this part of the code is quite convoluted and hard to work on. That is why I tried to clean it up a bit in this PR.

The main point of this PR was to solve the issue in the last two blocks. Any order defined for `responses`, `schemas` and `properties` is applied to the keys of their children. The parent objects themselves are never sorted alphabetically by their keys. That is why I introduced the `[*]` notation to make it clearer by which key (parent or self) we configure the order in (3).
The issue persists on current main.

Later, by @ed-randall-blk's request, I also added support for (1), which might have increased the scope of this PR a bit too much, I must admit. My array sorting also allows sorting by multiple fields, which again might be overkill.
(2) is not covered at all yet afaics.

What do yo think about introducing a completely new sorting configuration, besides the current one?
- when any of the existing two sorting configs exist, ignore the new one (maybe until the next major version)
- sensible defaults that sort almost everything, so the user will rarely have to define it
- just a single file instead of two
- auto-detect the config in the current workdir or npm project; no need to give an explicit parameter
- syntax that clearly shows what is being sorted and how, e.g. in the order (1) to (4) from above:
    ```json
    {
        "sortChildrenByAttribute": {
            "parameters": "name"
        },
        "sortChildrenByValue": ["enum", "required"],
        "sortKeysInOrder": {
            "info": ["title", "description", "version"],
            "schema": ["type", "items", "parameters"],
            "get": ["summary", "description", "operationId", "responses"]
        },
        "sortChildKeysInOrder": {
            "paths": ["get", "put", "patch", "post"],
            "parameters": ["name", "in", "description", "schema"],
            "schemas": ["type", "items", "parameters"]
        },
        "sortKeysAlphabetically": ["paths", "properties", "responses"]
    }
    ```

    `sortChildKeysInOrder` covers both the children of objects as well items in arrays.
- json-schema to validate that config file


_Sidenote:_
One other though was, that we might not have to be as generic in our sorting, but could only allow configuring orders for the exact types and attributes that occur in the openapi specification. This would also remove any redundancy e.g. between children of components.schemas and inline schema definitions, where we now have to define the key order two times.
On the other hand this would bind us very tightly to the details of the openapi spec and any change there could have implications for us.
The bigger concern would probably be assembling all the possible keys and types that could be configured instead of just going by arbitrary string keys.
That is why I still prefer the current approach.
